code/modules/xenoarchaeology/components/xenoartifact.dm

```diff
@@ -0,0 +1,410 @@
+
+/datum/component/xenoartifact/proc/build_traits(list/trait_list, amount, incompatabilities = TRUE)
+	var/list/options = trait_list
+	//Remove any blacklisted traits
+	options -= blacklisted_traits
+	//Remove any incompatible traits
+	if(incompatabilities)
+		options -= SSxenoarchaeology.get_trait_incompatibilities(parent)
+	for(var/i in 1 to amount)
+		//Pick a random trait
+		var/datum/xenoartifact_trait/T = pick_weight(options)
+		add_individual_trait(T)
+		options -= blacklisted_traits
+
+//Cooldown finish logic goes here
+/datum/component/xenoartifact/proc/reset_timer()
+	if(use_cooldown_timer)
+		deltimer(use_cooldown_timer)
+	use_cooldown_timer = null //Just incase
+
+/datum/component/xenoartifact/proc/get_extra_cooldowns()
+	var/time = 0 SECONDS
+	for(var/i in artifact_traits)
+		for(var/datum/xenoartifact_trait/T as() in artifact_traits[i])
+			time += T.cooldown
+	return time
+
+/datum/component/xenoartifact/proc/handle_malfunctions(itterate = TRUE)
+	//Instability rolls
+	if(!prob(instability))
+		if(itterate)
+			instability += artifact_type.instability_step
+		return
+	//Max malfunction checks, against our material
+	if(length(artifact_traits[TRAIT_PRIORITY_MALFUNCTION]) >= artifact_type.max_trait_malfunctions)
+		return
+	//Hint sound
+	var/atom/A = parent
+	playsound(A, 'sound/effects/light_flicker.ogg', 60)
+	A.visible_message("<span class='warning'>[A] makes a concerning sound, as if something has gone terribly wrong...</span>")
+	//Build malfunctions
+	var/list/focus_traits
+	focus_traits = SSxenoarchaeology.xenoa_malfunctions & artifact_type.get_trait_list()
+	build_traits(focus_traits, 1)
+	//Reset instability
+	instability = 0
+
+/datum/component/xenoartifact/proc/register_target(atom/target, force, type = XENOA_ACTIVATION_CONTACT)
+	//Don't register new targets unless the cooldown is finished
+	if((use_cooldown_timer || cooldown_override) && !force)
+		return
+	//Range check
+	if(get_dist(get_turf(parent), get_turf(target))> target_range && !force)
+		return
+	//Anti-artifact check
+	if(anti_check(target, type) && !force)
+		return
+	//Prexisting check
+	if((target in targets) && !force)
+		return
+	//Regular target follow through
+	create_beam(target)
+	targets += target
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, PROC_REF(unregister_target), TRUE)
+
+/datum/component/xenoartifact/proc/unregister_target(datum/source)
+	SIGNAL_HANDLER
+
+	targets -= source
+	UnregisterSignal(source, COMSIG_PARENT_QDELETING)
+
+/datum/component/xenoartifact/proc/on_examined(datum/source, mob/user, list/examine_text)
+	SIGNAL_HANDLER
+
+	var/mob/living/carbon/M = user
+	if(iscarbon(user) && M.can_see_reagents() || isobserver(user))
+		examine_text += "<span class='notice'>[parent] seems to be made from a [material_description]material.</span>"
+	//Special case for observers that shows all the traits
+	if(isobserver(user))
+		for(var/i in artifact_traits)
+			for(var/datum/xenoartifact_trait/T as() in artifact_traits[i])
```

Comment:
This list format is pretty horrendous, is there really no better way?

Reply:
This is the way I'm doing it